### ISSUE TITLE
Raster reprojection with triangulation and Proj4js

### ISSUE THREAD

**Issue by klokan** [openlayers/ol3#3785]
The modern web browsers provides technology allowing client-side raster reprojection of images on the fly - in JavaScript. OpenLayers v3 API was designed with the possibility to view raster geodata in a different map projection then the one served from a remote source (typically WMTS or WMS server, MapTiler tiles, etc).

This ticket presents a proposal (resulting from discussions at the Schladming code sprint) for an implementation of the on the fly transformation of raster data which should be directly integrated with the future version of the OpenLayers library.

The usage in API is going to be very simple:
```javascript
var map = new ol.Map({
  target: 'map',
  view: new ol.View({
    projection: 'EPSG:3857', // <---- HERE IS THE VIEW PROJECTION
    center: [0, 0],
    zoom: 2
  }),
  layers: [
    new ol.layer.Tile({
      source: new ol.source.TileWMS({
        projection: 'EPSG:4326', // <---- HERE IS THE DATA SOURCE PROJECTION
        url: 'http://demo.boundlessgeo.com/geoserver/wms',
        params: {
          'LAYERS': 'ne:NE1_HR_LC_SR_W_DR'
        }
    })
  })]
});
```

If a source (inheriting from `ol.source.TileImage` or `ol.source.Image`) has projection different from the current `ol.View`’s projection the reprojection happens automatically under the hood.

The overview for implementation of the reprojection:

- The reprojection process will be based on triangles -- the target raster will be divided into a limited number of triangles with vertices transformed using proj4js. The reprojection of pixels inside the triangle extent will be approximated with an affine transformation (with rendering hw-accelerated by the canvas 2d context):

![4d099f62-fe18-11e4-814d-da0c19a6f1b2](https://cloud.githubusercontent.com/assets/59284/8087649/81758cbc-0f9c-11e5-97ce-47706459678b.jpg)

- This way we can support wide range of projections from proj4js on almost any hardware and any web browser with canvas 2D support - without a need to have CORS headers on the remote server.
- The precision of the reprojection is limited by the number of triangles. To minimise the error a dynamic triangulation will be utilised:

![5a8c75ba-fe18-11e4-94a4-db173f6ff0ab](https://cloud.githubusercontent.com/assets/59284/8088179/bf394d42-0f9f-11e5-9bcb-dcfbc8694a56.png)

- In future, the above described initial implementation using Canvas2D can be replaced by a pixel perfect transformation utilising WebGL and a fragment shader with completely rewritten Proj4 equations into GLSL. See: [1](http://www.klokantech.com/labs/gpu-map-projection-warping-webgl/) [2](http://rjw57.github.io/proj4webgl/).

This proposal was discussed on the OL3-Dev Hangout on 2015-05-21.

The details for proposed OL3 integration are described in this design document: https://docs.google.com/document/d/1Vs3DTRjpgf53sqbBCfauGaPZa1iFX7Br8yDlEG_W6CI/edit?usp=sharing
Anybody is welcome the post comments.

@klokantech team started to work on this and will provide a pull request with a complete open-source implementation and test coverage latest by September 2015.

**Comment by bill-chadwick** [comment]
If you calculate an affine transform from 3 corners of the square (solving a simultaneous equation) then the maximum error will be at the centre of the square - where your red/green dots are. I assume you will be using two such transforms for each half of the square with triangular clipping regions. 

If you calculate an affine transform from 4 corners of a square (using least squares) then the maximum error is at the corners and the minimum error in the centre of the square. If you drive this error to be less than a threshold, then the stitching of tiles should also look good.

If the error threshold is exceeded then for the triangle method, aberrations will appear worst at the centre of a square tile and appear as a diagonal pattern. For the 4 point transform, aberrations will be worst where square tiles meet and appear in horizontal / vertical patterns (where there may already be a grid/graticule).

It would be interesting to look at the performance tradeoff between using two 3 point transforms and triangular clipping as opposed to one 4 point transforms with simple rectangular rendering.

To deal with composition (overlaying or blending of multiple layers) then it is perhaps the ol.View that should hold the master resolutions table with the re-projection transforms including a re-scale from the nearest tile source resolution to the view resolution. Expending is generally better than shrinking.

I expect that trying to reproject e.g. a Web Mercator world view to say Mollweide would result in too many transforms. So you might want to limit the number of transforms as well as specifying the accuracy.

**Comment by bill-chadwick** [comment]
Oh and ol has recently been updated to support non square tiles - that should not be a problem for this technique.

**Comment by jacmendt** [comment]
Beside the automatic reprojection of an image according to his projection property, it would be nice if there is away to explicit define a transformation matrix (translate, scale, rotate) of an ol.source.Source.

**Comment by bill-chadwick** [comment]
The images below show the error between the affine transformation and proj4s transformation, from EPSG:27700 to EPSG:3857, for a 5km tile of 20m pixels near the middle of the British Grid. The first image is the error from two triangular transforms, each calculated from three corners. The second image is for a single transform calculated from all four corners. In both cases, the worst error (Red) is around 0.07 pixels (or 1.4 m).

![two_triangles](https://cloud.githubusercontent.com/assets/10521516/8113679/932eb6a6-1067-11e5-9ad2-5d3c2e93056d.PNG)
![square](https://cloud.githubusercontent.com/assets/10521516/8113688/966e1fbe-1067-11e5-8d26-3120e850210b.PNG)

**Comment by marcjansen** [comment]
I really hope @bill-chadwick and @klokantech sort this out... Soooooo looking forward to this

**Comment by petrsloup** [comment]
@bill-chadwick Thank you for the comment and nice visualizations of the error distribution. You are correct that calculating the transformation for triangles results in highest error being in the middle (although we are calculating two transformations instead of one, so the average error should be slightly lower).

The main advantage of this approach is that the vertices are transformed precisely and the triangle edges are basically just a linear interpolations of the vertices. Because of this, the edges always fit perfectly (even between tiles). If there is for example grid/graticule present in the source data, it will be reprojected as polyline using the triangle-based approach, but if we went for the quad-based approach, there could actually be discontinuities on the tile edges -- the segment ends would not meet.

Overall, I think it should not matter where exactly the worst error will be, as long as it's not visually distracting -- that's why we went for the triangle-based approach.

**Comment by bill-chadwick** [comment]
I expect that at street map level one affine transform will often suffice
for the whole view. Whether squares or triangles are used, as long as the
subdivision reduces the maximum error to less than half a pixel then the
results should look good.

**Comment by bill-chadwick** [comment]
I am still not convinced that triangular transform tiling will give best performance. I think that the rendering of two clipped triangles is almost bound to be slower than the rendering of one rectangle of the same area.

With a rectangular 4 point derived transform, the corner coordinates will already be available in both coordinate systems to check the affine transform's accuracy. If subdivision is required, as per the second of the red+green+blue dot figures above,  five new points need to be reprojected. 

With two triangle based transforms, four coordinates will have been reprojected to construct them but a fifth point must be reprojected to do the accuracy check at the center. If subdivision is required, eight new points must be reprojected (the previous center conversion can be reused).

My experiments for UK Grid and Swiss Grid to Web Mercator using square tiles of the same size show similar maximum errors for a four point transform or two triangular transforms. The maximum errors in the two rainbow figures above are the same.

**Comment by klokan** [comment]
PR with implementation is in #4122

**Comment by marcjansen** [comment]
Closed with #4122
